### PR TITLE
objstore: Added common acceptance test for common objstore behaviour.

### PR DESCRIPTION
**Pull request by Bplotka**
Additionally:
- Fixed inmen storage.
- Added leaktest where we can.
- Added ForeachObjStore test suite.

Currently tested by CI: GCS (only for contributors with write perm) and inmem only

Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: pkg/objstore/gcs/gcs.go [hunk 1]
```diff
@@ -32,6 +38,8 @@ type Bucket struct {
 	opsTotal *prometheus.CounterVec
 }
 
+var _ objstore.Bucket = &Bucket{}
```
Comment (stefan-improbable) [inline pkg/objstore/gcs/gcs.go]: Can you add a comment on why is this here?

File: pkg/objstore/inmem/inmem.go [hunk 2]
```diff
@@ -34,18 +39,40 @@ func (b *Bucket) Objects() map[string][]byte {
...
+		parts := strings.SplitAfter(filename, objstore.DirDelim)
+		unique[strings.Join(parts[:dirPartsCount+1], "")] = struct{}{}
 	}
+
 	var keys []string
 	for n := range unique {
 		keys = append(keys, n)
 	}
-	sort.Strings(keys)
+	sort.Slice(keys, func(i, j int) bool {
```
Comment (stefan-improbable) [inline pkg/objstore/inmem/inmem.go]: Why do we need this ordering here? Also, why can't we just use normal sort?

File: pkg/objstore/inmem/inmem.go [hunk 3]
```diff
@@ -57,23 +84,31 @@ func (b *Bucket) Iter(_ context.Context, dir string, f func(string) error) error
...
 // Get returns a reader for the given object name.
 func (b *Bucket) Get(_ context.Context, name string) (io.ReadCloser, error) {
+	if name == "" {
+		return nil, errors.New("inmem: object name is empty")
+	}
+
 	file, ok := b.objects[name]
 	if !ok {
-		return nil, errors.Errorf("no such file %s", name)
+		return nil, errNotFound
```
Comment (stefan-improbable) [inline pkg/objstore/inmem/inmem.go]: I like having filename inside the error message :), can we keep it?

File: pkg/objstore/inmem/inmem.go [hunk 4]
```diff
@@ -57,23 +84,31 @@ func (b *Bucket) Iter(_ context.Context, dir string, f func(string) error) error
...
 	}
 
 	return ioutil.NopCloser(bytes.NewReader(file)), nil
 }
 
 // GetRange returns a new range reader for the given object name and range.
 func (b *Bucket) GetRange(_ context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	if name == "" {
+		return nil, errors.New("inmem: object name is empty")
```
Comment (stefan-improbable) [inline pkg/objstore/inmem/inmem.go]: can you add name to this error wessage?

File: pkg/objstore/inmem/inmem.go [hunk 5]
```diff
@@ -105,3 +140,8 @@ func (b *Bucket) Delete(_ context.Context, name string) error {
 	delete(b.objects, name)
 	return nil
 }
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	return err == errNotFound
```
Comment (stefan-improbable) [inline pkg/objstore/inmem/inmem.go]: You can create a new type based on error and do a type assertion

File: pkg/objstore/objtesting/foreach.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	"github.com/improbable-eng/thanos/pkg/objstore/gcs"
+	"github.com/improbable-eng/thanos/pkg/objstore/inmem"
+	"github.com/improbable-eng/thanos/pkg/objstore/s3"
+	"github.com/improbable-eng/thanos/pkg/testutil"
+)
+
+// ForeachStore runs given test using all available objstore implementations.
+// For each it creates a new bucket with a random name and a cleanup function
+// that deletes it after test was run.
+func ForeachStore(t *testing.T, testFn func(t testing.TB, bkt objstore.Bucket)) {
```
Comment (stefan-improbable) [inline pkg/objstore/objtesting/foreach.go]: Is this also copypasted? If no, I don't think we should be using env variables. We should have instead:
`pkg/compact/e2e_test.go` exporting fn `testE2E(t *testing.T, s Store)`
`pkg/objtesting/acceptance_test.go` exporting fn `testAcceptance(t *testing.T, s Store)`
`foreach.go` exporting `testAll`:
```
func testAll(t *testing.T, s Store) {
  - run e2e test as subtest
  - run acceptance test as subtest
  ...
}
```

and call testAll from s3/ package, gcs package, etc.

File: pkg/objstore/inmem/inmem.go (hunk 2 above)
Comment (Bplotka) [reply pkg/objstore/inmem/inmem.go]: To match ordering of GCS. I decided to took  that as reference. We can choose our own ordering, but we would end up changing order for GCS code instead. The thing is that Thanos works well with GCS so did not want to change ordering (:

File: pkg/objstore/inmem/inmem.go (hunk 3 above)
Comment (Bplotka) [reply pkg/objstore/inmem/inmem.go]: I can, sure, just wanted to keep it simple (I need to assert not found err in tests). If not I would endup doing some `IsNotFoundErr` magic. Can do if you want.

File: pkg/objstore/inmem/inmem.go (hunk 4 above)
Comment (Bplotka) [reply pkg/objstore/inmem/inmem.go]: well, it is empty so not sure what should I add xD

File: pkg/objstore/inmem/inmem.go (hunk 5 above)
Comment (Bplotka) [reply pkg/objstore/inmem/inmem.go]: yea, we can just it is more complex. Will do.

File: pkg/objstore/objtesting/foreach.go (hunk 6 above)
Comment (Bplotka) [reply pkg/objstore/objtesting/foreach.go]: No, it is new logic, let's talk offline, maybe your idea is better.

File: pkg/objstore/objstore.go [hunk 7]
```diff
@@ -275,6 +279,10 @@ func (b *metricBucket) Delete(ctx context.Context, name string) error {
 	return err
 }
 
+func (b *metricBucket) IsObjNotFoundErr(err error) bool {
+	return b.IsObjNotFoundErr(err)
```
Comment (domgreen) [inline pkg/objstore/objstore.go]: Little confused here ... does this not just call itself?

Comment (domgreen) [reply pkg/objstore/objstore.go]: should it be `b.bkt.IsObjNotFoundErr(err)`?

File: pkg/objstore/gcs/gcs.go (hunk 1 above)
Comment (domgreen) [reply pkg/objstore/gcs/gcs.go]: I find it a bit confusing that its initialised if its not used or needed for a specific side effect. Especially if all types of buckets are covered by tests.

File: pkg/objstore/s3/s3.go [hunk 8]
```diff
@@ -202,3 +205,57 @@ func (b *Bucket) Delete(ctx context.Context, name string) error {
...
 }
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	return minio.ToErrorResponse(err).Code == "NoSuchKey"
+}
+
+func configFromEnv() *Config {
+	c := &Config{
+		Bucket:    os.Getenv("S3_BUCKET"),
```
Comment (domgreen) [inline pkg/objstore/s3/s3.go]: Is there anyway we can get around using environment vars in tests?

File: pkg/objstore/gcs/gcs.go (hunk 1 above)
Comment (Bplotka) [reply pkg/objstore/gcs/gcs.go]: makes sense.

File: pkg/objstore/inmem/inmem.go (hunk 3 above)
Comment (Bplotka) [reply pkg/objstore/inmem/inmem.go]: especially it is inmem ONLY used in tests.. let's just leave it simple, what do you think?

File: pkg/objstore/inmem/inmem.go (hunk 5 above)
Comment (Bplotka) [reply pkg/objstore/inmem/inmem.go]: actually `inmem` is only used in tests, so I would vote for keeping it simple.

File: pkg/objstore/objstore.go (hunk 7 above)
Comment (Bplotka) [reply pkg/objstore/objstore.go]: yup, nice catch, thanks!

File: pkg/objstore/s3/s3.go (hunk 8 above)
Comment (Bplotka) [reply pkg/objstore/s3/s3.go]: not really ): We need custom parameters to have tests against real s3 right?

File: pkg/objstore/inmem/inmem.go (hunk 3 above)
Comment (stefan-improbable) [reply pkg/objstore/inmem/inmem.go]: sgtm
